src/api/worker/rest/EntityRestCache.ts

```diff
@@ -255,110 +243,186 @@ export class EntityRestCache implements IEntityRestCache {
+
+			await this.updateRangeInStorage(typeRef, listId, count, reverse, entities)
+
+			// Try to get enough entities from cache
+			const entitiesFromCache = await this.storage.provideFromRange(typeRef, listId, start, count, reverse)
+
+			// If we exhausted the entities from the server,
+			// Or cache was able to provide the original request
+			// Then we are done
+			if (entities.length < count || entitiesFromCache.length === count) {
+				break
+			}
+		}
+	}
 
-					// put the new elements into the cache
-					await this.handleElementRangeResult(typeRef, listId, loadStartId, count, reverse, entities, count)
+	/**
+	 * Loads all elements from the startId in the direction of the range
+	 * Once complete, returns as many elements as it can from the original request
+	 * range:          |--------------|
+	 * request:                             <------*
+	 * range becomes:  |---------------------------|
+	 * or
+	 * range:                   |---------|
+	 * request:            <-------------------*
+	 * range becomes:      |--------------------|
+	 */
+	private async extendTowardsRange<T extends ListElementEntity>(typeRef: TypeRef<T>, listId: Id, start: Id, count: number, reverse: boolean) {
 
-					// provide from cache with the actual start id
-					const resultElements = await this.storage.provideFromRange(typeRef, listId, start, count, reverse)
+		while (true) {
 
-					if (entities.length < count || resultElements.length === count) {
-						// either all available elements have been loaded from target or the requested number of elements could be provided from cache
-						return resultElements
-					} else {
-						// try again with the new elements in the cache
-						return this.loadRange(typeRef, listId, start, count, reverse)
-					}
-				} else {
-					// Start is outside the range, and we are loading towards the range, so grow outward from the range until start is
-					// inside the range, then provide from cache
+			const range = assertNotNull(await this.storage.getRangeForList(typeRef, listId))
 
-					const loadStartId = reverse
-						? range.upper
-						: range.lower
+			const loadStartId = reverse
+				? range.upper
+				: range.lower
 
-					const entities = await this.entityRestClient.loadRange(typeRef, listId, loadStartId, count, !reverse)
+			const entities = await this.entityRestClient.loadRange(typeRef, listId, loadStartId, count, !reverse)
 
-					await this.handleElementRangeResult(typeRef, listId, loadStartId, count, !reverse, entities, count)
+			await this.updateRangeInStorage(typeRef, listId, count, !reverse, entities)
 
-					if (!await this.storage.isElementIdInCacheRange(typeRef, listId, start)) {
-						return this.loadRange(typeRef, listId, start, count, reverse)
-					} else {
-						return this.storage.provideFromRange(typeRef, listId, start, count, reverse)
-					}
-				}
+			if (await this.storage.isElementIdInCacheRange(typeRef, listId, start)) {
```

Comment:
Is this enough? Can we get stuck here? Should we check that we get less than we asked?

Reply:
`updateRangeInStorage` checks that and will set the range to `GENERATED_MIN_ID`/`GENERATED_MAX_ID` in this case, so it is implied

Reply:
ah, so it will set it to min/max and then this check will succeed? Let's leave a comment then